src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,309 @@
+                        "OAUTHBEARERAUTH",
+                        "SASL OAUTHBEARER authentication failed (principal=%s): %s",
+                        state->md_principal_name,
+                        state->server_error_msg.ptr);
+                r = -1;
+                break;
+        }
+
+        if (out.ptr) {
+                r = rd_kafka_sasl_send(rktrans, out.ptr, (int)out.size,
+                                       errstr, errstr_size);
+                rd_free(out.ptr);
+        }
+
+        return r;
+}
+
+
+/**
+ * @brief Handle received frame from broker.
+ */
+static int rd_kafka_sasl_oauthbearer_recv (rd_kafka_transport_t *rktrans,
+                                     const void *buf, size_t size,
+                                     char *errstr, size_t errstr_size) {
+        const rd_chariov_t in = { .ptr = (char *)buf, .size = size };
+        return rd_kafka_sasl_oauthbearer_fsm(rktrans, &in, errstr, errstr_size);
+}
+
+
+/**
+ * @brief Initialize and start SASL OAUTHBEARER (builtin) authentication.
+ *
+ * Returns 0 on successful init and -1 on error.
+ *
+ * @locality broker thread
+ */
+static int rd_kafka_sasl_oauthbearer_client_new (rd_kafka_transport_t *rktrans,
+                                    const char *hostname,
+                                    char *errstr, size_t errstr_size) {
+        struct rd_kafka_sasl_oauthbearer_state *state;
+        state = rd_calloc(1, sizeof(*state));
+        state->state = RD_KAFKA_SASL_OAUTHBEARER_STATE_SEND_CLIENT_FIRST_MESSAGE;
+        /*
+         * Save off the state structure now, before any possibility of
+         * returning, so that we will always free up the allocated memory in
+         * rd_kafka_sasl_oauthbearer_close().
+         */
+        rktrans->rktrans_sasl.state = state;
+        /*
+         * Make sure we have a consistent view of the token and extensions
+         * throughout the authentication process -- even if it is refreshed
+         * midway through this particular authentication.
+         */
+        rwlock_rdlock(&rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->refresh_lock);
+        if (!rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->token_value) {
+                rwlock_rdunlock(&rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->refresh_lock);
+                rd_snprintf(errstr, errstr_size, "OUTHBEARER cannot log in because there is no token available");
+                return -1;
+        }
+        state->token_value = strdup(rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->token_value);
+        state->md_principal_name = strdup(rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->md_principal_name);
+        rd_list_copy_to(&state->extensions, &rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->extensions,
+                rd_strtup_list_copy, NULL);
+        rwlock_rdunlock(&rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->refresh_lock);
+
+        /* Confirm there is no explicit "auth" extension */
```

Comment:
Isn't this already checked in set_token() ?

Reply:
Yes!  Removed from here.